---
title: "omicsR2"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{omicsR2}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

R package to investigate how much variance of a given trait is explained by
the joined effects of different omics (e.g. Methylome),
centered around the `greml` function described in the `qgg` R package

omicsR2 fits a GREML model in which one or more omics are defined as
random effects and uses cross-validation to estimate explained variance.

1. The first step is to compute a similarity matrix based on
genotype/DNA methylation levels/expression levels. 
2. In a second step, the similarity.
matrix is then included as random effect predicting an outcome in a 
training sample (e.g. random 80% of sample).
3. Based on the results of the training sample, predicted estimates (GBLUP)
are computed in validation sample.
4. The correlation between predicted and observed values is used to estimate
the variance explained.
5. Variance explained by a baseline model with covariates only is substracted
from full model.
6. Steps 2-5 are repeated to obtain a distribution of cross-validated variance
explained estimates across random training/validation splits.

For details, see description for 
[qgg::greml](https://psoerensen.github.io/qgg/articles/GBLUP_cross_validation.html)

qgg reference:
Palle Duun Rohde, Izel Fourie Sørensen, Peter Sørensen,
qgg: an R package for large-scale quantitative genetic analyses,
Bioinformatics, Volume 36, Issue 8, 15 April 2020, Pages 2614–2615,
https://doi.org/10.1093/bioinformatics/btz955


## Example

Below you find an example analysis with simulated data. In this example we aim
to estimate the variance explained in an outcome by DNA methylation beyond 
the variance explained by covariates. 

```{r example}
# Load BGData for similarity matrix computation
library(BGData)

# Load omicsR2 package for GREML analysis
library(omicsR2)

# Load methylation and phenotype data
data("cpg_simulated") # Methylation data
data("phenotype") # Simulated phenotypes data
data("Batch")     # Batch similarity matrix

# The methylation data contains methylation levels
# for 1000 CpG probes (columns)
# and 500 participants (rows)
cpg_simulated[1:5,1:5]

# The phenotype data contains simulated outcome data and two covariates
head(phenotype, 5)

# Also a batch matrix is included, defining which participants are on the same
# batch (1)
Batch[c(1:5,100:105),c(1:5,100:105)]

# Compute similarity matrix for DNA methylation
# getG does the following:
# 1. Z-score standardizes the methylation values
# 2. Takes the product of the transpose
# 3. Scales the similarity matrix so mean diagonal is 1
Gmt <- getG(cpg_simulated)
Gmt[1:5, 1:5]

# Match phenotype and methylation similarity matrix by ID variable,
# so that participants appear in the same order
matched.list <- match_pheno_similarity(phenotype, Gmt, "ID")

# Save the data as separate objects again
phenotype_matched <- matched.list[[1]]
Gmt_matched <- matched.list[[2]]

# Regress outcome on methylation similarity matrix, batch similarity matrix,
# two fixed effect covariates with GREML and Monte-Carlo Cross-Validation.
# Returns variance explained by DNA methylation minus the variance explained by
# fixed effect covariates and similarity matrices specified in random_baseline.
# Training::validation ratio is 80:20 and randomly sampled 100 times.
Gmt_variance_explained <- omicsR2(outcome = "outcome",
                                  fixed_covar = "covariate1 + covariate2",
 random_full = list(Methylation=Gmt_matched, Batch=Batch),
  random_baseline = list(Batch=Batch),
 data = phenotype_matched,
 validation_proportion = 0.2, repetitions = 100, seed = 20190405)

# Examine the variance explained distribution obtained from cross-validation
head(Gmt_variance_explained)
mean(Gmt_variance_explained)
quantile(Gmt_variance_explained)
hist(Gmt_variance_explained, breaks = 10)
```
